draft-ietf-tls-esni.md

```diff
@@ -569,42 +566,58 @@ implementations need to take care to ensure that sensitive extensions are not
-In particular, this is possible because the "pre_shared_key" extension is
-forbidden in ClientHelloOuter.
+### Encrypting the ClientHello {#encrypting-clienthello}
 
-The client then generates the HPKE encryption context and computes the
-encapsulated key, context, and payload as:
+To construct the "encrypted_client_hello", the client first determines the
+encapsulated key and HPKE encryption context. If constructing the first
+ClientHelloOuter, it computes them as:
 
 ~~~
     pkR = DeserializePublicKey(ECHConfig.contents.public_key)
     enc, context = SetupBaseS(pkR,
                               "tls ech" || 0x00 || ECHConfig)
-    payload = context.Seal(ClientHelloOuterAAD,
-                           EncodedClientHelloInner)
 ~~~
 
-Note that the HPKE functions DeserializePublicKey and SetupBaseS are those which
-match `ECHConfig.contents.kem_id` and the AEAD/KDF used with `context` are those
-which match the client's chosen preference from
-`ECHConfig.contents.cipher_suites`.  The `info` parameter to SetupBaseS is the
-concatenation of "tls ech", a zero byte, and the serialized ECHConfig.
+If constructing the second ClientHelloOuter ({{client-hrr}}), it reuses the
+encryption context computed for the first ClientHelloOuter, and sets `enc` to
+the empty string. Note that the HPKE context maintains a sequence number, so
+this operation internally uses a fresh nonce for each AEAD operation. Reusing
+the HPKE context avoids an attack described in {{flow-hrr-hijack}}.
 
-The value of the "encrypted_client_hello" extension in the ClientHelloOuter is
-a `ClientECH` with the following values:
+The client then computes ClientHelloOuterAAD ({{authenticating-outer}}) by
+constructing a ClientHello with all other extensions determined as in
+{{real-ech}}.
+
+Next, the client determines the length L of encrypting EncodedClientHelloInner
+with the selected HPKE AEAD. This is typically the sum of the plaintext length
+and the AEAD tag length. The client fills in a "encrypted_client_hello"
+extension with the following values:
 
 - `config_id`, the identifier corresponding to the chosen ECHConfig structure;
 - `cipher_suite`, the client's chosen cipher suite;
 - `enc`, as computed above; and
-- `payload`, as computed above.
+- `payload`, a placeholder byte string containing L zeros.
 
 If optional configuration identifiers (see {{optional-configs}}) are used,
-`config_id` SHOULD be set to a randomly generated byte. Unless specified by the
-application using (D)TLS or externally configured on both sides,
-implementations MUST set the field as specified in
-{{encrypted-client-hello}}.
+`config_id` SHOULD be set to a randomly generated byte in the first
+ClientHelloOuter and MUST be left unchanged for the second ClientHelloOuter.
+
+The client serializes this structure compute the ClientHelloOuterAAD. It then
+computes the payload as:
+
+~~~
+    payload = context.Seal(ClientHelloOuterAAD,
```

Comment:
Maybe we call this enc_payload (or final_payload or something) to distinguish between the payload extension value? The following sentence read a bit weird to me, where it refers to "payload" and "this value."

Reply:
Done

Reply:
(Went with final_payload because enc_payload seemed potentially confusing with enc.)

Reply:
👍